src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -174,63 +174,127 @@ private void MakeRep(int type, int min, int max)
         }
 
         /// <summary>Performs additional optimizations on an entire tree prior to being used.</summary>
+        /// <remarks>
+        /// Some optimizations are performed by the parser while parsing, and others are performed
+        /// as nodes are being added to the tree.  The optimizations here expect the tree to be fully
+        /// formed, as they inspect relationships between nodes that may not have been in place as
+        /// individual nodes were being processed/added to the tree.
+        /// </remarks>
         internal RegexNode FinalOptimize()
         {
             RegexNode rootNode = this;
-            Debug.Assert(rootNode.Type == Capture && rootNode.ChildCount() == 1);
+            Debug.Assert(rootNode.Type == Capture);
+            Debug.Assert(rootNode.Next is null);
+            Debug.Assert(rootNode.ChildCount() == 1);
 
-            // If we find backtracking construct at the end of the regex, we can instead make it non-backtracking,
-            // since nothing would ever backtrack into it anyway.  Doing this then makes the construct available
-            // to implementations that don't support backtracking.
-            if ((Options & RegexOptions.RightToLeft) == 0 && // only apply optimization when LTR to avoid needing additional code for the rarer RTL case
-                (Options & RegexOptions.Compiled) != 0) // only apply when we're compiling, as that's the only time it would make a meaningful difference
+            if ((Options & RegexOptions.RightToLeft) == 0) // only apply optimization when LTR to avoid needing additional code for the rarer RTL case
             {
-                // Walk the tree, starting from the sole child of the root implicit capture.
-                RegexNode node = rootNode.Child(0);
-                while (true)
+                // Optimization: backtracking removal at expression end.
+                // If we find backtracking construct at the end of the regex, we can instead make it non-backtracking,
+                // since nothing would ever backtrack into it anyway.  Doing this then makes the construct available
+                // to implementations that don't support backtracking.
+                if ((Options & RegexOptions.Compiled) != 0) // only apply when we're compiling, as that's the only time it would make a meaningful difference
```

Comment:
>  as that's the only time it would make a meaningful difference

Don't we cache even if we don't compile? If so wouldn't this work here be potentially amortized?

Reply:
> Don't we cache even if we don't compile?

We don't cache / use the cache from the ctors, e.g. this doesn't touch the cache:
```C#
new Regex(...).IsMatch()
```
We only touch the cache from the static methods.

That said, even if it was ammortized, my thinking was it doesn't buy a whole lot for the interpreter.  This affects whether the node is backtracked into, but there's nothing to backtrack into for the last node in the expression.  As such, it doesn't actually save anything for the interpreter, other than having to interpret one or two additional commands to jump past the relevant entries in the back stack.  For the compiler, there are a couple of benefits.  The biggest is that if everything else in the expression is backtracking-free, we then upgrade to that special compile path that generates much nicer / more efficient code.